### DISCUSSION TITLE
IKEv2 VPN has stopped working with macOS Ventura

### DISCUSSION THREAD

**Discussion by xXConcasXx** [strongswan/strongswan#1377] [Q&A] [answered]
I upgraded to macOS Ventura and my VPN has stopped working, no longer connects and no UI-level error messages appear.
I also tried a clean install but nothing has changed, I assume something has changed from the previous iOS version.

From Console application I tried to log while trying to connect by filtering system.log with keyword 'ikev2' and this is the result:

[macos_log.txt](https://github.com/strongswan/strongswan/files/9872921/macos_log.txt)

Reading the log these messages caught my attention:

> errore	20:33:45.956428+0200	NEIKEv2Provider	Bootstrapping; external subsystem UIKit_PKSubsystem refused setup
> errore	20:33:45.966253+0200	NEIKEv2Provider	open flag(s) 0x01000000 are reserved for VFS use and do not affect behaviour when passed to sqlite3_open_v2
> errore	20:33:45.966312+0200	NEIKEv2Provider	cannot open file at line 46922 of [6bf7a27121]
> errore	20:33:45.966331+0200	NEIKEv2Provider	os_unix.c:46922: (2) open(/private/var/db/DetachedSignatures) - No such file or directory
> errore	20:33:46.050318+0200	NEIKEv2Provider	[IKE_SA_INIT R resp0 48DFB37B75AC41B5-0000000000000000] Initiator init received notify error Error Domain=NEIKEv2ProtocolErrorDomain Code=14 "NoProposalChosen" UserInfo={NSDebugDescription=NoProposalChosen}
> errore	20:33:46.050409+0200	NEIKEv2Provider	IKEv2Session[1, 48DFB37B75AC41B5-0000000000000000] Failed to process IKE SA Init packet (connect)
> errore	20:33:46.061254+0200	NEIKEv2Provider	Failed to find suitable address, path supports IPv4 yes IPv6 no

Also I captured the logs from the server by reading **/var/log/syslog** and this is the result:

> Oct 26 20:49:53 raspberrypi charon: 06[NET] received packet: from x.x.x.x[500] to 192.168.83.3[500] (432 bytes)
> Oct 26 20:49:53 raspberrypi charon: 06[ENC] parsed IKE_SA_INIT request 0 [ SA KE No N(REDIR_SUP) N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) ]
> Oct 26 20:49:53 raspberrypi charon: 06[IKE] x.x.x.x is initiating an IKE_SA
> Oct 26 20:49:53 raspberrypi charon: 06[IKE] local host is behind NAT, sending keep alives
> Oct 26 20:49:53 raspberrypi charon: 06[IKE] remote host is behind NAT
> Oct 26 20:49:53 raspberrypi charon: 06[IKE] received proposals unacceptable
> Oct 26 20:49:53 raspberrypi charon: 06[ENC] generating IKE_SA_INIT response 0 [ N(NO_PROP) ]
> Oct 26 20:49:53 raspberrypi charon: 06[NET] sending packet: from 192.168.83.3[500] to x.x.x.x[500] (36 bytes)

My server Strongswan configuration is this:

`/etc/ipsec.conf`

```
config setup
    charondebug="ike 1, knl 1, cfg 0"
    uniqueids=no

conn user
    auto=add
    compress=no
    type=tunnel
    keyexchange=ikev2
    fragmentation=yes
    forceencaps=yes
    dpdaction=clear
    dpddelay=300s
    rekey=no
    left=%any
    leftid=@vpn.*****.***
    leftcert=server-cert.pem
    leftsendcert=always
    leftsubnet=0.0.0.0/0
    right=%any
    rightid=%any
    rightauth=eap-mschapv2
    rightsourceip=10.10.10.0/24
    rightdns=1.1.1.1,1.0.0.1
    rightsendcert=never
    eap_identity=%identity
    ike=chacha20poly1305-sha512-curve25519-prfsha512,aes256gcm16-sha384-prfsha384-ecp384,aes256-sha1-modp1024,aes128-sha1-modp1024,3des-sha1-modp1024!
    esp=chacha20poly1305-sha512,aes256gcm16-ecp384,aes256-sha256,aes256-sha1,3des-sha1!
```

`/etc/ipsec.secrets`

```
# This file holds shared secrets or RSA private keys for authentication.

# RSA private key for this host, authenticating it to any other host
# which knows the public part.

: RSA "server-key.pem"
user : EAP "&iC3OdvqAKRJbvSy18@IDBpvu0Kee*@S"
```

`/etc/ufw/before.rules`

```
#
# rules.before
#
# Rules that should be run before the ufw command line added rules. Custom
# rules should be added to one of these chains:
#   ufw-before-input
#   ufw-before-output
#   ufw-before-forward
#

# Don't delete these required lines, otherwise there will be errors

*nat
-A POSTROUTING -s 10.10.10.0/24 -o eth0 -m policy --pol ipsec --dir out -j ACCEPT
-A POSTROUTING -s 10.10.10.0/24 -o eth0 -j MASQUERADE
COMMIT

*mangle
-A FORWARD --match policy --pol ipsec --dir in -s 10.10.10.0/24 -o eth0 -p tcp -m tcp --tcp-flags SYN,RST SYN -m tcpmss --mss 1361:1536 -j TCPMSS --set-mss 1360
COMMIT

*filter
:ufw-before-input - [0:0]
:ufw-before-output - [0:0]
:ufw-before-forward - [0:0]
:ufw-not-local - [0:0]
# End required lines

-A ufw-before-forward --match policy --pol ipsec --dir in --proto esp -s 10.10.10.0/24 -j ACCEPT
-A ufw-before-forward --match policy --pol ipsec --dir out --proto esp -d 10.10.10.0/24 -j ACCEPT

# allow all on loopback
-A ufw-before-input -i lo -j ACCEPT
-A ufw-before-output -o lo -j ACCEPT

# quickly process packets for which we already have a connection
-A ufw-before-input -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A ufw-before-output -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A ufw-before-forward -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT

# drop INVALID packets (logs these in loglevel medium and higher)
-A ufw-before-input -m conntrack --ctstate INVALID -j ufw-logging-deny
-A ufw-before-input -m conntrack --ctstate INVALID -j DROP

# ok icmp codes for INPUT
-A ufw-before-input -p icmp --icmp-type destination-unreachable -j ACCEPT
-A ufw-before-input -p icmp --icmp-type time-exceeded -j ACCEPT
-A ufw-before-input -p icmp --icmp-type parameter-problem -j ACCEPT
-A ufw-before-input -p icmp --icmp-type echo-request -j ACCEPT

# ok icmp code for FORWARD
-A ufw-before-forward -p icmp --icmp-type destination-unreachable -j ACCEPT
-A ufw-before-forward -p icmp --icmp-type time-exceeded -j ACCEPT
-A ufw-before-forward -p icmp --icmp-type parameter-problem -j ACCEPT
-A ufw-before-forward -p icmp --icmp-type echo-request -j ACCEPT

# allow dhcp client to work
-A ufw-before-input -p udp --sport 67 --dport 68 -j ACCEPT

#
# ufw-not-local
#
-A ufw-before-input -j ufw-not-local

# if LOCAL, RETURN
-A ufw-not-local -m addrtype --dst-type LOCAL -j RETURN

# if MULTICAST, RETURN
-A ufw-not-local -m addrtype --dst-type MULTICAST -j RETURN

# if BROADCAST, RETURN
-A ufw-not-local -m addrtype --dst-type BROADCAST -j RETURN

# all other non-local packets are dropped
-A ufw-not-local -m limit --limit 3/min --limit-burst 10 -j ufw-logging-deny
-A ufw-not-local -j DROP

# allow MULTICAST mDNS for service discovery (be sure the MULTICAST line above
# is uncommented)
-A ufw-before-input -p udp -d 224.0.0.251 --dport 5353 -j ACCEPT

# allow MULTICAST UPnP for service discovery (be sure the MULTICAST line above
# is uncommented)
-A ufw-before-input -p udp -d 239.255.255.250 --dport 1900 -j ACCEPT

# don't delete the 'COMMIT' line or these rules won't be processed
COMMIT
```
`/etc/ufw/sysctl.conf`

```
#
# Configuration file for setting network variables. Please note these settings
# override /etc/sysctl.conf and /etc/sysctl.d. If you prefer to use
# /etc/sysctl.conf, please adjust IPT_SYSCTL in /etc/default/ufw. See
# Documentation/networking/ip-sysctl.txt in the kernel source code for more
# information.
#

# Uncomment this to allow this host to route packets between interfaces
#net/ipv4/ip_forward=1
#net/ipv6/conf/default/forwarding=1
#net/ipv6/conf/all/forwarding=1

# Disable ICMP redirects. ICMP redirects are rarely used but can be used in
# MITM (man-in-the-middle) attacks. Disabling ICMP may disrupt legitimate
# traffic to those sites.

# Ignore bogus ICMP errors
net/ipv4/icmp_echo_ignore_broadcasts=1
net/ipv4/icmp_ignore_bogus_error_responses=1
net/ipv4/icmp_echo_ignore_all=0

# Don't log Martian Packets (impossible addresses)
# packets
net/ipv4/conf/all/log_martians=0
net/ipv4/conf/default/log_martians=0

#net/ipv4/tcp_fin_timeout=30
#net/ipv4/tcp_keepalive_intvl=1800

# Uncomment this to turn off ipv6 autoconfiguration
#net/ipv6/conf/default/autoconf=1
#net/ipv6/conf/all/autoconf=1

# Uncomment this to enable ipv6 privacy addressing
#net/ipv6/conf/default/use_tempaddr=2
#net/ipv6/conf/all/use_tempaddr=2
net/ipv4/ip_forward=1
net/ipv4/conf/all/accept_redirects=0
net/ipv4/conf/all/send_redirects=0

net/ipv4/conf/all/accept_redirects=0
net/ipv4/conf/default/accept_redirects=0
net/ipv4/conf/all/send_redirects=0
net/ipv6/conf/all/accept_redirects=0
net/ipv6/conf/default/accept_redirects=0

net/ipv4/ip_no_pmtu_disc=1
```

**Comment by xXConcasXx** [comment, marked as answer]
Setting a more detailed log level I noticed this:

> IKE_SA (unnamed)[4] state change: CREATED => CONNECTING
> Oct 26 21:36:29 raspberrypi charon: 10[CFG] selecting proposal:
> Oct 26 21:36:29 raspberrypi charon: 10[CFG]   no acceptable ENCRYPTION_ALGORITHM found
> Oct 26 21:36:29 raspberrypi charon: 10[CFG] selecting proposal:
> Oct 26 21:36:29 raspberrypi charon: 10[CFG]   no acceptable ENCRYPTION_ALGORITHM found
> Oct 26 21:36:29 raspberrypi charon: 10[CFG] selecting proposal:
> Oct 26 21:36:29 raspberrypi charon: 10[CFG]   no acceptable INTEGRITY_ALGORITHM found
> Oct 26 21:36:29 raspberrypi charon: 10[CFG] selecting proposal:
> Oct 26 21:36:29 raspberrypi charon: 10[CFG]   no acceptable ENCRYPTION_ALGORITHM found
> Oct 26 21:36:29 raspberrypi charon: 10[CFG] selecting proposal:
> Oct 26 21:36:29 raspberrypi charon: 10[CFG]   no acceptable ENCRYPTION_ALGORITHM found
> Oct 26 21:36:29 raspberrypi charon: 10[CFG] received proposals: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_2048
> Oct 26 21:36:29 raspberrypi charon: 10[CFG] configured proposals: IKE:CHACHA20_POLY1305/PRF_HMAC_SHA2_512/CURVE_25519, IKE:AES_GCM_16_256/PRF_HMAC_SHA2_384/ECP_384, IKE:AES_CBC_256/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_1024, IKE:AES_CBC_128/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_1024, IKE:3DES_CBC/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_1024

On ipsec.conf I've these:

```
ike=chacha20poly1305-sha512-curve25519-prfsha512,aes256gcm16-sha384-prfsha384-ecp384,aes256-sha1-modp1024,aes128-sha1-modp1024,3des-sha1-modp1024!
esp=chacha20poly1305-sha512,aes256gcm16-ecp384,aes256-sha256,aes256-sha1,3des-sha1!
```

I apologize for the silly and trivial question, but any of you can explain to me how to add the proposed algorithms without distorting the correct syntax of the parameters?

**Reply by Thermi** [reply]
just add aes256-sha256-prfsha256-modp2048 to the list for ike
(the prfsha256 is redundant with the sha256 because sha256 also implicitely specifies the prf)

**Reply by xXConcasXx** [reply]
> basta aggiungere aes256-sha256-prfsha256-modp2048 all'elenco per ike (il prfsha256 è ridondante con lo sha256 perché sha256 specifica implicitamente anche il prf)

Thank you!

**Reply by flinhong** [reply]
This fixed issue on macOS, but failed for Windows client 😢

**Reply by dareal68** [reply]
This fixed the issue for me (MacBook Pro 2023, Ventura).
Thank you very much

**Reply by chris1012T** [reply]
@Thermi 
This fixed the issue for me (MacBook Pro 2017, Ventura; Mac mini 2023, Sonoma; iPhone 15 Pro, iOS 17.1.1).
Thank you very much!

**Reply by d00ker** [reply]
This is quite interesting, but this solution does not work for me on Sonoma. The default values ​​esp and ike from the official strongswan docs work fine.
```
ike=aes128-sha256-modp3072
esp=aes128-sha256
```

**Reply by memorynoise** [reply]
Started experiencing the same problem after I upgraded to iOS 26, I can confirm that @xXConcasXx 's solution fixed my connection issues using strongSwan 5.8.2 on ubuntu 20.04.6.

Weirdly enough I couldn't replicate it on my macbook m1 running mac os Tahoe, which was still working fine even after the update.

**Comment by infrabro** [comment]
I recently had the same problem on my VPN server. Thermi's answer helped me to solve the issue.  Here's my config + Thermi's fix. I think my original config was the same as OP's config. Tested on MacOS High Sierra ,Big Sur, Monterey, Ventura and Windows 10/11.


Add/replace to /etc/ipsec.conf under conn
`ike=chacha20poly1305-sha512-curve25519-prfsha512,aes256-sha256-prfsha256-modp2048,aes256gcm16-sha384-prfsha384-ecp384,aes256-sha1-modp1024,aes128-sha1-modp1024,3des-sha1-modp1024!`
`esp=chacha20poly1305-sha512,aes256gcm16-ecp384,aes256-sha256,aes256-sha1,3des-sha1!`

**Reply by flinhong** [reply]
Thanks a bunch, your suits work on macOS and Windows now 👏.

...but still freezed on Android 12, no matter strongSwan client nor native IKEv2 support.

```
parsed IKE_AUTH request 1 [ IDi IDr SA TSi TSr CPRQ(ADDR ADDR6 DNS DNS6 MASK VER) ]

deleting half open IKE_SA with [xxx.xxx.xxx.xxx] after timeout
```

**Reply by Thermi** [reply]
Need logs from the strongSwan client

**Reply by Yraidu** [reply]
Thanks a lot, I confirm it works on Macos Ventura, Windows 11, Android 13 and ios 16

**Reply by ccchhh1** [reply]
> Thanks a lot, I confirm it works on Macos Ventura, Windows 11, Android 12 and ios 16

Hello, I did not find this file, I use macos ventura, how to edit this file, thank you

**Reply by dareal68** [reply]
/etc/ipsec.conf is server side not client side (Ventura)

**Reply by ccchhh1** [reply]
> /etc/ipsec.conf is server side not client side (Ventura)

thanks

**Reply by MrAlonso** [reply]
Thank you very much, it keeps working on MacOS Ventura 13.0 (but not on MacOS Ventura 13.4.1, so be careful with that), also on Windows 11, Android 13 and iOS 16 as well

**Comment by knarou34** [comment]
Hello, 
Same problem with mac OS ventura can't connect on PFSENSE IKEv2 
But i need to keep Windows 10,11 and MacOS 11, 12 and now 13
if you see log in pfsense ipsec, you see : 
```
received proposals: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_2048
Jun 1 18:22:34	charon	83915	09[CFG] <61665> 
configured proposals: 
IKE:AES_GCM_16_256/PRF_HMAC_SHA1/MODP_2048, 
IKE:AES_CBC_256/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_2048, 
IKE:3DES_CBC/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_2048, 
IKE:3DES_CBC/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_1024
```
So, i  add some parameters in Phase 1 and Phase 2 IPsec in PFSENSE
**On Phase 1**
I add algorithm 
Algo => KeyLenght=>Hash=>DH Group
`------------------------------------------------`
AES => 256 bits => SHA256 => 14(2048bit)

**On Phase 2 Proposal**
I add
Encryption Algorithms => AES256-GCM
Hash Algorithms => SHA256
PFS key group => 14(2048bit)

And normaly, it's OK to connect you MacOS Ventura to VPN IKEv2
And it still work for Windows 10 too
I hope it will be useful.

**Comment by xg590** [comment]
For those coming here late, these ike and sep cipher suites works well for my Windows 10, MacOS Ventura 13.4 and iOS 16.5.
* /etc/ipsec.conf
```
conn foo123
    ......
    keyexchange=ikev2
    ......
    ike=aes256-sha256-modp1024,aes256-sha1-modp1024,3des-sha1-modp1024
    esp=aes256-sha256,aes256-sha1,3des-sha1
```

**Reply by tobiasbrunner** [reply]
I'd definitely recommend using a stronger key exchange method than `modp1024`. At least `modp2048`, which is supported by all platforms you list (see [the docs for help with Windows clients](https://docs.strongswan.org/docs/5.9/interop/windowsClients.html)).
